src/native/libs/System.Security.Cryptography.Native/pal_evp_mac.c

```diff
@@ -0,0 +1,383 @@
+    if (API_EXISTS(EVP_MAC_init))
+    {
+        assert(API_EXISTS(OSSL_PARAM_construct_octet_string));
+        assert(API_EXISTS(OSSL_PARAM_construct_end));
+
+        ERR_clear_error();
+
+        size_t keyLengthT = Int32ToSizeT(keyLength);
+
+        OSSL_PARAM params[4] = { 0 };
+        int i = 0;
+        params[i++] = OSSL_PARAM_construct_octet_string(OSSL_MAC_PARAM_KEY, (void*) key, keyLengthT);
+        params[i++] = OSSL_PARAM_construct_int32(OSSL_MAC_PARAM_XOF, &xof);
+
+        if (customizationString && customizationStringLength > 0)
+        {
+            size_t customizationStringLengthT = Int32ToSizeT(customizationStringLength);
+            params[i++] = OSSL_PARAM_construct_octet_string(OSSL_MAC_PARAM_CUSTOM, (void*) customizationString, customizationStringLengthT);
+        }
+
+        params[i] = OSSL_PARAM_construct_end();
+
+        if (!EVP_MAC_init(ctx, NULL, 0, params))
+        {
+            return 0;
+        }
+
+        return 1;
+    }
+#else
+    (void)ctx;
+    (void)key;
+    (void)keyLength;
+    (void)customizationString;
+    (void)customizationStringLength;
+    (void)xof;
+#endif
+
+    assert(0 && "Inconsistent EVP_MAC API availability.");
+    return -2;
+}
+
+int32_t CryptoNative_EvpMacReset(EVP_MAC_CTX* ctx)
+{
+    if (ctx == NULL)
+    {
+        return -1;
+    }
+
+#ifdef NEED_OPENSSL_3_0
+    if (API_EXISTS(EVP_MAC_init))
+    {
+        ERR_clear_error();
+
+        if (!EVP_MAC_init(ctx, NULL, 0, NULL))
+        {
+            return 0;
+        }
+
+        return 1;
+    }
+#endif
+
+    assert(0 && "Inconsistent EVP_MAC API availability.");
+    return -2;
+}
+
+int32_t CryptoNative_EvpMacUpdate(EVP_MAC_CTX* ctx, uint8_t* data, int32_t dataLength)
+{
+    if (ctx == NULL || (data == NULL && dataLength > 0) || dataLength < 0)
+    {
+        return -1;
+    }
+
+#ifdef NEED_OPENSSL_3_0
+    if (API_EXISTS(EVP_MAC_update))
+    {
+        ERR_clear_error();
+
+        if (dataLength > 0)
+        {
+            size_t dataLengthT = Int32ToSizeT(dataLength);
+
+            if (!EVP_MAC_update(ctx, data, dataLengthT))
+            {
+                return 0;
+            }
+        }
+
+        return 1;
+    }
+#endif
+
+    assert(0 && "Inconsistent EVP_MAC API availability.");
+    return -2;
+}
+
+int32_t CryptoNative_EvpMacFinal(EVP_MAC_CTX* ctx, uint8_t* mac, int32_t macLength)
+{
+    if (ctx == NULL || (mac == NULL && macLength > 0) || macLength < 0)
+    {
+        return -1;
+    }
+
+#ifdef NEED_OPENSSL_3_0
+    if (API_EXISTS(EVP_MAC_final))
+    {
+        assert(API_EXISTS(OSSL_PARAM_construct_end));
+        assert(API_EXISTS(OSSL_PARAM_construct_int32));
+        assert(API_EXISTS(EVP_MAC_CTX_set_params));
+
+        ERR_clear_error();
+
+        OSSL_PARAM params[2] = { 0 };
+
+        size_t macLengthT = Int32ToSizeT(macLength);
+        params[0] = OSSL_PARAM_construct_int32(OSSL_MAC_PARAM_SIZE, &macLength);
+        params[1] = OSSL_PARAM_construct_end();
```

Comment:
Since there are no conditionals, couldn't this be

```suggestion
        size_t macLengthT = Int32ToSizeT(macLength);

        OSSL_PARAM params[2] =
        {
            OSSL_PARAM_construct_int32(OSSL_MAC_PARAM_SIZE, &macLength),
            OSSL_PARAM_construct_end()
        };
```

maybe even without the 2?